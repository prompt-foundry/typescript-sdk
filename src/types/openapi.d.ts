import type {
  OpenAPIClient,
  Parameters,
  UnknownParamsObject,
  OperationResponse,
  AxiosRequestConfig,
} from 'openapi-client-axios';

declare namespace Components {
    namespace Schemas {
        export interface Error {
            /**
             * example:
             * Prompt ID is required.
             */
            error?: string;
        }
        export interface PromptConfiguration {
            /**
             * example:
             * PROMPT_1
             */
            promptId: string;
            promptParameters: {
                /**
                 * example:
                 * 0
                 */
                frequencyPenalty: number;
                /**
                 * example:
                 * gpt-3.5-turbo
                 */
                modelName: string;
                /**
                 * example:
                 * 0
                 */
                presencePenalty: number;
                /**
                 * example:
                 * TEXT
                 */
                responseFormat: "TEXT" | "JSON";
                /**
                 * example:
                 * 1
                 */
                temperature: number;
                /**
                 * example:
                 * 1
                 */
                topP: number;
                /**
                 * example:
                 * 100
                 */
                maxTokens: number | null;
                /**
                 * example:
                 * 97946543
                 */
                seed: number | null;
                /**
                 * example:
                 * checkWeather
                 */
                toolChoice?: string | null;
            };
            promptMessages: /* The initial messages to be included with your call to the LLM API. */ PromptMessage[];
            promptTools: Tool[];
        }
        /**
         * The initial messages to be included with your call to the LLM API.
         */
        export interface PromptMessage {
            /**
             * example:
             * Hello, {{city}}!
             */
            content: string | null;
            /**
             * example:
             * USER
             */
            role: "USER" | "ASSISTANT" | "SYSTEM" | "TOOL";
            /**
             * example:
             * TOOL_CALL_1
             */
            toolCallId: string | null;
            toolCalls: ToolFunctionCall[] | null;
        }
        export interface Tool {
            /**
             * example:
             * TOOL_1
             */
            toolId?: string;
            /**
             * The name of the tool to be called. Must be a-z, A-Z, 0-9, or contain underscores and dashes, with a maximum length of 64.
             */
            name: string; // ^[a-zA-Z0-9_-]{1,64}$
            /**
             * A description of what the tool does, used by the model to choose when and how to call the tool.
             */
            description: string;
            parameters: /* The parameters the functions accepts, described as a JSON Schema object. This schema is designed to match the TypeScript Record<string, unknown>, allowing for any properties with values of any type. */ ToolParameters;
        }
        export interface ToolFunctionCall {
            /**
             * example:
             * TOOL_CALL_1
             */
            toolCallId: string;
            /**
             * The type of the tool. Currently, only `function` is supported.
             * example:
             * function
             */
            type: "function";
            function: {
                /**
                 * The arguments to call the function with, as generated by the model in JSON format. Note that the model does not always generate valid JSON, and may hallucinate parameters not defined by your function schema. Validate the arguments in your code before calling your function.
                 * example:
                 * {}
                 */
                arguments: string;
                /**
                 * The name of the function to call.
                 * example:
                 * checkWeather
                 */
                name: string;
            };
        }
        /**
         * The parameters the functions accepts, described as a JSON Schema object. This schema is designed to match the TypeScript Record<string, unknown>, allowing for any properties with values of any type.
         */
        export interface ToolParameters {
            [name: string]: any;
        }
    }
}
declare namespace Paths {
    namespace SdkV1Prompts$PromptId {
        namespace Get {
            export interface HeaderParameters {
                "X-API-KEY": Parameters.XAPIKEY;
            }
            namespace Parameters {
                export type PromptId = string;
                export type XAPIKEY = string;
            }
            export interface PathParameters {
                promptId: Parameters.PromptId;
            }
            namespace Responses {
                export type $200 = Components.Schemas.PromptConfiguration;
                export type $400 = Components.Schemas.Error;
                export type $401 = Components.Schemas.Error;
                export type $404 = Components.Schemas.Error;
                export type $500 = Components.Schemas.Error;
            }
        }
    }
}

export interface OperationMethods {
}

export interface PathsDictionary {
  ['/sdk/v1/prompts/{promptId}']: {
  }
}

export type Client = OpenAPIClient<OperationMethods, PathsDictionary>
