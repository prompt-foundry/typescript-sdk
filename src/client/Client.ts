import { BASE_URL } from '../constants'
import { api, createApiClient } from '../openapi/client'

import EvaluationAssertions from './EvaluationAssertions'
import Evaluations from './Evaluations'
import Prompts from './Prompts'
import Tools from './Tools'

export default class PromptFoundry {
  private client: typeof api

  prompts: Prompts

  tools: Tools

  evaluations: Evaluations

  evaluationAssertions: EvaluationAssertions

  constructor({ apiKey, baseUrl = BASE_URL }: { apiKey: string; baseUrl?: string }) {
    // @ts-expect-error - protect against missing apiKey for js consumers
    if (apiKey === '' || apiKey === undefined || apiKey === null || apiKey === 0) {
      throw new Error('Prompt Foundry API Key is required!')
    }

    this.client = createApiClient(baseUrl)
    this.prompts = new Prompts({ client: this.client, apiKey })
    this.tools = new Tools({ client: this.client, apiKey })
    this.evaluations = new Evaluations({ client: this.client, apiKey })
    this.evaluationAssertions = new EvaluationAssertions({ client: this.client, apiKey })
  }
}
