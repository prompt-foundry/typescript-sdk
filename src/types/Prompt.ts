export type responseFormat = {
  type: 'text' | 'json_object'
}

export type PromptParameters = {
  frequencyPenalty: number
  maxTokens?: number
  modelName: string
  presencePenalty: number
  responseFormat: responseFormat
  seed?: number
  temperature: number
  topP: number
}

export enum PromptMessageRole {
  USER = 'user',
  ASSISTANT = 'assistant',
  TOOL = 'tool',
  SYSTEM = 'system'
}

export type PromptMessage = {
  content: string
  role: PromptMessageRole
}

export type Prompt = {
  promptId: string
  promptParameters: PromptParameters
  promptMessages: PromptMessage[]
}
