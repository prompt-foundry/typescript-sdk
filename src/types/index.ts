import { Components } from './openapi'

export type PromptConfiguration = Components.Schemas.PromptConfiguration
export type PromptMessage = Components.Schemas.PromptConfiguration['promptMessages'][0]
