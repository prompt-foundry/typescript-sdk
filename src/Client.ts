import { ChatCompletionCreateParamsNonStreaming } from 'openai/resources'

import APIClient from './ApiClient'
import { getMissingPromptVariables, renderPromptWithVariables, validatePromptVariables } from './helpers'
import { mapPromptToOpenAIConfig } from './helpers/openAi'
import { PromptConfiguration } from './types'

export default class PromptFoundry {
  private client: APIClient

  constructor({ apiKey }: { apiKey: string }) {
    // @ts-expect-error - protect against missing apiKey for js consumers
    if (apiKey === '' || apiKey === undefined || apiKey === null || apiKey === 0) {
      throw new Error('Prompt Foundry API Key is required!')
    }

    this.client = new APIClient({
      baseURL: 'https://api.promptfoundry.ai/sdk/v1',
      apiToken: apiKey
    })
  }

  public async getRawPrompt({ id }: { id: string }): Promise<PromptConfiguration> {
    return this.client.get<PromptConfiguration>(`/prompts/${id}`)
  }

  public async getPrompt({ id, variables }: { id: string; variables: Record<string, string> }): Promise<PromptConfiguration> {
    const result = await this.getRawPrompt({ id })

    if (!validatePromptVariables(result, variables)) {
      const missingVariables = getMissingPromptVariables(result, variables)
      throw new Error(`Missing variables in prompt: ${missingVariables.join(', ')}`)
    }
    return renderPromptWithVariables(result, variables)
  }

  public async getOpenAiPrompt({
    id,
    variables
  }: {
    id: string
    variables: Record<string, string>
  }): Promise<ChatCompletionCreateParamsNonStreaming> {
    const updatedWithVariables = await this.getPrompt({ id, variables })

    return mapPromptToOpenAIConfig(updatedWithVariables)
  }
}
