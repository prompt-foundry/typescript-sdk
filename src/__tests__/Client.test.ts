import APIClient from '../ApiClient'
import Client from '../Client'
import { PromptConfiguration, PromptMessageRoleEnum } from '../types'

describe('client', () => {
  describe('init client', () => {
    test('inits', () => {
      expect(() => new Client({ apiKey: 'HELLO_WORLD_API_KEY' })).not.toThrow()
    })

    test('throws if no api key - empty string', () => {
      expect(() => new Client({ apiKey: '' })).toThrow()
    })

    test('throws if no api key - no args', () => {
      // @ts-expect-error - testing for no args
      expect(() => new Client()).toThrow()
    })
  })
  describe('getPrompt', () => {
    it('returns prompt', async () => {
      const client = new Client({ apiKey: '123' })

      const value = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: {
            type: 'text'
          },
          temperature: 0.7,
          topP: 1
        },
        messages: [
          {
            content: 'Hello, world!',
            role: 'user'
          },
          {
            content: 'Hi there!',
            role: 'assistant'
          }
        ]
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      const prompt = await client.getPrompt({ id: 'HELLO', variables: {} })

      expect(prompt).toEqual(value)
    })

    it('should replace variables', async () => {
      const client = new Client({ apiKey: '123' })

      const value: PromptConfiguration = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: 'TEXT',
          temperature: 0.7,
          topP: 1,
          seed: null,
          maxTokens: null,
          toolChoice: 'auto'
        },
        messages: [
          {
            content: 'Hello, world!',
            role: PromptMessageRoleEnum.USER,
            toolCalls: null,
            toolCallId: null
          },
          {
            content: 'Hi there {{name}}!',
            role: PromptMessageRoleEnum.ASSISTANT,
            toolCalls: null,
            toolCallId: null
          }
        ],
        tools: []
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      const prompt = await client.getPrompt({ id: 'HELLO', variables: { name: 'bob' } })

      expect(prompt.messages[1].content).toEqual('Hi there bob!')
    })

    it('should throw error if missing variables', async () => {
      const client = new Client({ apiKey: '123' })

      const value: PromptConfiguration = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: 'TEXT',
          temperature: 0.7,
          topP: 1,
          seed: null,
          maxTokens: null,
          toolChoice: 'auto'
        },
        messages: [
          {
            content: 'Hello, world!',
            role: PromptMessageRoleEnum.USER,
            toolCalls: null,
            toolCallId: null
          },
          {
            content: 'Hi there {{name}}!',
            role: PromptMessageRoleEnum.ASSISTANT,
            toolCalls: null,
            toolCallId: null
          }
        ],
        tools: []
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      await expect(() => client.getPrompt({ id: 'HELLO', variables: {} })).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  describe('getOpenAiPrompt', () => {
    it('returnsmapped prompt', async () => {
      const client = new Client({ apiKey: '123' })

      const value: PromptConfiguration = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: 'TEXT',
          temperature: 0.7,
          topP: 1,
          seed: null,
          maxTokens: null,
          toolChoice: 'auto'
        },
        messages: [
          {
            content: 'Hello, world!',
            role: PromptMessageRoleEnum.USER,
            toolCalls: null,
            toolCallId: null
          },
          {
            content: 'Hi there!',
            role: PromptMessageRoleEnum.ASSISTANT,
            toolCalls: null,
            toolCallId: null
          }
        ],
        tools: []
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      const prompt = await client.getOpenAiPrompt({ id: 'HELLO', variables: {} })

      expect(prompt).toMatchSnapshot()
    })

    it('should replace variables', async () => {
      const client = new Client({ apiKey: '123' })

      const value: PromptConfiguration = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: 'TEXT',
          temperature: 0.7,
          topP: 1,
          seed: null,
          maxTokens: null,
          toolChoice: 'auto'
        },
        messages: [
          {
            content: 'Hello, world!',
            role: PromptMessageRoleEnum.USER,
            toolCalls: null,
            toolCallId: null
          },
          {
            content: 'Hi there {{name}}!',
            role: PromptMessageRoleEnum.ASSISTANT,
            toolCalls: null,
            toolCallId: null
          }
        ],
        tools: []
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      const prompt = await client.getOpenAiPrompt({ id: 'HELLO', variables: { name: 'bob' } })

      expect(prompt.messages[1].content).toEqual('Hi there bob!')
    })

    it('should throw error if missing variables', async () => {
      const client = new Client({ apiKey: '123' })

      const value: PromptConfiguration = {
        id: 'HELLO',
        name: 'winning',
        parameters: {
          frequencyPenalty: 0,
          modelName: 'gpt-3.5-turbo',
          presencePenalty: 0,
          responseFormat: 'TEXT',
          temperature: 0.7,
          topP: 1,
          seed: null,
          maxTokens: null,
          toolChoice: 'auto'
        },
        messages: [
          {
            content: 'Hello, world!',
            role: PromptMessageRoleEnum.USER,
            toolCalls: null,
            toolCallId: null
          },
          {
            content: 'Hi there {{name}}!',
            role: PromptMessageRoleEnum.ASSISTANT,
            toolCalls: null,
            toolCallId: null
          }
        ],
        tools: []
      }

      const apiClient: APIClient = {
        // @ts-expect-error - mocking the client's client property
        axiosInstance: {}, // Mocked as an empty object or with specific mock functions if needed
        get: jest.fn().mockResolvedValue(value)
      }

      // @ts-expect-error - mocking the client's client property
      client.client = apiClient

      await expect(() => client.getOpenAiPrompt({ id: 'HELLO', variables: {} })).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
