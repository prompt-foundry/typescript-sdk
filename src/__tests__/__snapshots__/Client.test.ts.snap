// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client getOpenAiPrompt returnsmapped prompt 1`] = `
{
  "frequency_penalty": 0,
  "max_tokens": null,
  "messages": [
    {
      "content": "Hello, world!",
      "name": undefined,
      "role": "user",
    },
    {
      "content": "Hi there!",
      "name": undefined,
      "role": "assistant",
      "tool_calls": undefined,
    },
  ],
  "model": "gpt-3.5-turbo",
  "presence_penalty": 0,
  "response_format": {
    "type": "text",
  },
  "seed": null,
  "temperature": 0.7,
  "tool_choice": "auto",
  "tools": [],
  "top_p": 1,
}
`;

exports[`client getOpenAiPrompt should throw error if missing variables 1`] = `"Missing variables in prompt: name"`;

exports[`client getPrompt should throw error if missing variables 1`] = `"Missing variables in prompt: name"`;
